Library/ViewModels/PostCampaignCheckoutViewModel.swift

```diff
@@ -335,7 +343,57 @@ public class PostCampaignCheckoutViewModel: PostCampaignCheckoutViewModelType,
 
     // MARK: CompleteOnSessionCheckout
 
-    // TODO: Call .completeOnSessionCheckout when self.confirmPaymentSuccessfulProperty.signal is called
+    let completeCheckoutWithExistingCardInput: Signal<GraphAPI.CompleteOnSessionCheckoutInput, Never> = Signal
+      .combineLatest(self.confirmPaymentSuccessfulProperty.signal.skipNil(), checkoutId, selectedExistingCard)
+      .map { (
+        clientSecret: String,
+        checkoutId: String,
+        selectedCard: (source: PaymentSourceSelected, paymentMethodId: String, isNewPaymentMethod: Bool)?
+      ) -> GraphAPI.CompleteOnSessionCheckoutInput in
+
+      GraphAPI
+        .CompleteOnSessionCheckoutInput(
+          checkoutId: encodeToBase64("Checkout-\(checkoutId)"),
+          paymentIntentClientSecret: clientSecret,
+          paymentSourceId: selectedCard?.paymentMethodId,
+          paymentSourceReusable: true,
+          applePay: nil
+        )
+      }
+
+    let completeCheckoutWithNewCardInput: Signal<GraphAPI.CompleteOnSessionCheckoutInput, Never> = Signal
+      .combineLatest(self.confirmPaymentSuccessfulProperty.signal.skipNil(), checkoutId, selectedNewCard)
+      .map { (
+        clientSecret: String,
+        checkoutId: String,
+        _: (source: PaymentSourceSelected, paymentMethodId: String, isNewPaymentMethod: Bool)?
+      ) -> GraphAPI.CompleteOnSessionCheckoutInput in
+
+      GraphAPI
+        .CompleteOnSessionCheckoutInput(
+          checkoutId: encodeToBase64("Checkout-\(checkoutId)"),
+          paymentIntentClientSecret: clientSecret,
+          paymentSourceId: nil,
+          paymentSourceReusable: true,
+          applePay: nil
+        )
+      }
+
+    // TODO: Implement ApplePay
+    // let completeCheckoutWithApplePay =
+
+    let checkoutCompleteSignal = Signal
+      .merge(
+        completeCheckoutWithExistingCardInput,
+        completeCheckoutWithNewCardInput
+        // completeCheckoutWithApplePayInput
+      )
+      .switchMap { input in
+        AppEnvironment.current.apiService.completeOnSessionCheckout(input: input).materialize()
+      }
+
+    self.checkoutComplete = checkoutCompleteSignal.signal.ignoreValues()
```

Comment:
nit: I'm not sure about this, but I think this might need to be `checkoutCompleteSignal.signal.values().ignoreValues()` in order to prevent it from also firing if there's an error. Will you test it if you haven't already?

Reply:
Ha, yup, you're right. Happened to me a few times.